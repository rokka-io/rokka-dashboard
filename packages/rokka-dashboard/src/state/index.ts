import { checkAuthentication, fetchOperations, fetchStackOptions, OperationsResponse, StackOptionsResponse } from '../api';
import { set as setCookie } from '../utils/cookie';

export const SESSION_COOKIE_KEY = 'rka_session';

export interface AppUser {
  organization: string;
  apiKey: string;
}

export interface AppState {
  showSidebar: boolean;
  user?: AppUser;
  operations?: OperationsResponse;
  stackOptions?: StackOptionsResponse;
}

const defaultState: AppState = {
  showSidebar: false
};

let internalState = defaultState;

type Subscriber = (state: AppState) => void;

let subscribers: Subscriber[] = [];

/**
 * Subscribe callback to state updates.
 */
export function subscribe(callback: Subscriber) {
  subscribers.push(callback);
}

/**
 * Unsubscribe existing callback from state updates.
 */
export function unsubscribe(callback: Subscriber) {
  subscribers = subscribers.filter(cb => cb !== callback);
}

/**
 * Merge internalState with updated partialStates.
 */
function updateState(...partialStates: Array<Partial<AppState>>) {
  internalState = Object.assign({}, internalState, ...partialStates);
  notifySubscribers();
}

/**
 * Notify subscribers with current state.
 */
function notifySubscribers() {
  if (subscribers.length) {
    for (const subscriber of subscribers) {
      if (subscriber) {
        subscriber(internalState);
      }
    }
  }
}

/**
 * Checks if organization/apiKey combination is correct and logs the user in.
 */
export async function login(organization: string, apiKey: string) {
  const authenticated = await checkAuthentication(organization, apiKey);

  if (authenticated === true) {
    const user = { organization, apiKey };
    updateState({ user });
    updateState(await listOperations(), await getStackOptions());
    setCookie(SESSION_COOKIE_KEY, { user });
    return Promise.resolve();
  }

  setCookie(SESSION_COOKIE_KEY, {}); // clear session on error

  return Promise.reject(new Error('Invalid authentication'));
}

/**
 * Fetches stack operations.
 */
async function listOperations(): Promise<Partial<AppState>> {
  try {
    const operations = await fetchOperations();
    return { operations };
  } catch (err) {
    return {};
  }
}

/**
 * Fetches stack options.
 */
async function getStackOptions(): Promise<Partial<AppState>> {
  try {
    const stackOptions = await fetchStackOptions();
    return { stackOptions };
  } catch (err) {
    return {};
  }
}

/**
 * Show or hide sidebar.
 */
export function toggleSidebar() {
  updateState({ showSidebar: !internalState.showSidebar });
}

/**
 * Export current state (readonly).
 */
export const currentState = () => Object.assign({}, internalState);

/**
 * Do not use. Only used for tests.
 */
export const testResetState = () => {
  internalState = defaultState;
};

export const testSubscriberCount = () => subscribers.length;
