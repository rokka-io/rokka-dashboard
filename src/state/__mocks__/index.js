export function listStacks () {
  throw new Error('foo')
  return {}
}
