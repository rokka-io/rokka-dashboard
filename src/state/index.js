import rokka, { authenticate, resetClient } from '../rokka'
import { get as getCookie, set as setCookie, del as delCookie } from '../utils/cookie'

const SESSION_COOKIE_KEY = 'rka_session'

const session = getCookie(SESSION_COOKIE_KEY)

if (session && session.auth) {
  const { auth } = session
  login(auth.organization, auth.apiKey)
}

const defaultState = {
  showSidebar: false,
  stackClone: {},
  auth: null,
  uploadedImages: [],
  deletedImages: [],
  stacks: {},
  operations: {},
  stackOptions: null,
  alert: null
}

let internalState = defaultState

let listener = null

/**
 * Update state and notify listener.
 *
 * @param {*} partialState
 */
function updateState(partialState) {
  internalState = Object.assign({}, internalState, partialState)
  if (!listener && process.env.NODE_ENV !== 'production') {
    console.warn('No listener added to state updates.')
    return
  }
  listener(internalState)
}

/**
 * Show or hide sidebar.
 */
export function toggleSidebar() {
  updateState({ showSidebar: !internalState.showSidebar })
}

/**
 * Delete an image.
 *
 * @param {string} hash
 *
 * @return {Promise}
 */
export function deleteImage(hash) {
  return rokka()
    .sourceimages.delete(internalState.auth.organization, hash)
    .then(() => {
      updateState({
        deletedImages: [...internalState.deletedImages, hash]
      })
      clearUploadedImages()
    })
}

export function clearDeletedImages() {
  updateState({ deletedImages: [] })
}

export function updateUploadedImages(uploadedImages) {
  updateState({ uploadedImages })
  // In case user reuploads a just deleted image.
  clearDeletedImages()
}

export function clearUploadedImages() {
  updateState({ uploadedImages: [] })
}

/**
 * Login sets the API key to the rokka API client.
 *
 * @param {string}   organization
 * @param {string}   apiKey
 * @param {Function} successCb    Success callback
 *
 * @returns {Promise}
 */
export function login(organization, apiKey, successCb) {
  authenticate(apiKey)

  return rokka()
    .organizations.get(organization)
    .then(() => {
      // remove alert in case there was auth failed before.
      removeAlert()

      const done = () => {
        updateState({ auth: { organization, apiKey } })
        setCookie(
          SESSION_COOKIE_KEY,
          { auth: { organization, apiKey } },
          { maxAge: 3600 * 18, secure: window.location.protocol === 'https:' }
        )
        listOperations()
        getDefaultStackOptions()
      }

      if (!successCb) {
        return done()
      }

      successCb(done)
    })
    .catch(err => {
      console.error(err)

      setCookie(SESSION_COOKIE_KEY, {}) // clear session on error

      if (err.statusCode === 403 || err.statusCode === 404) {
        updateState({ auth: null })
        setAlert('error', 'Authentication failed')
      } else if (err.statusCode === 400) {
        updateState({ auth: null })
        setAlert('error', 'Organization name is not valid')
      } else if (err.statusCode === 429) {
        updateState({ auth: null })
        setAlert('error', 'Too many requests')
      }
      setAlert('error', 'Error getting organizations', 10000)
      throw err
    })
}

/**
 * Logout clears the store and removes the API key from rokka API client.
 * Additionally it also removes the cookie.
 */
export function logout() {
  resetClient()
  delCookie(SESSION_COOKIE_KEY)
  updateState(defaultState)
}

/**
 * List stacks of an organization and populate the global state with the result.
 *
 * @param {number} limit
 *
 * @returns {Promise}
 */
export function listStacks() {
  const { items = [], filter = '' } = internalState.stacks

  return rokka()
    .stacks.list(internalState.auth.organization)
    .then(({ body }) => {
      sortAlphabetically(body.items)
      const newItems = [...items, ...body.items]
      updateState({
        stacks: {
          items: newItems,
          filter,
          filteredItems: newItems.filter(item => filterItem(item, filter))
        }
      })
    })
    .catch(err => {
      console.error(err)
      if (err.statusCode === 403) {
        updateState({ auth: null })
        setAlert('error', 'Authentication failed')
        return
      }
      setAlert('error', 'Error when fetching stacks', 10000)
    })
}

function filterItem(item, filter) {
  return item.name.indexOf(filter) > -1
}

export function filterStacks(filter) {
  const items = internalState.stacks.items.filter(item => filterItem(item, filter))
  updateState({
    stacks: { ...internalState.stacks, filter, filteredItems: items }
  })
}

/**
 * Resets the internal state with stacks and fetches them again.
 *
 * TODO: Slightly ugly that we have to manually override the internalState.stacks.
 *
 * @returns {Promise}
 */
export function refreshStacks() {
  internalState.stacks = { ...internalState.stacks, items: [], total: 0, currentOffset: 0 }
  return listStacks()
}

/**
 * Sorts Alphabetically
 *
 * @param {array} items
 *
 * @returns {array}
 */
function sortAlphabetically(items) {
  return items.sort((a, b) => {
    if (a.name < b.name) {
      return -1
    }
    if (a.name > b.name) {
      return 1
    }
    return 0
  })
}

/**
 * List operations and populate the global state with the result.
 */
export function listOperations() {
  rokka()
    .operations.list()
    .then(({ body }) => {
      updateState({
        operations: body
      })
    })
    .catch(err => {
      console.error(err)
      if (err.statusCode === 403) {
        updateState({ auth: null })
        setAlert('error', 'Authentication failed')
        return
      }
      setAlert('error', 'Error when fetching operations', 10000)
    })
}

/**
 * Fetch default stack options and populate the global state with the result.
 */
export function getDefaultStackOptions() {
  rokka()
    .stackoptions.get()
    .then(({ body }) => {
      updateState({
        stackOptions: body
      })
    })
    .catch(err => {
      console.error(err)
      setAlert('error', 'Error fetching default stack options', 10000)
    })
}

/**
 * Create a new stack
 *
 * @param {string}      name
 * @param {Array}       operations
 * @param {Object|null} options
 *
 * @return {Promise}
 */
export function createStack(name, operations, options) {
  return rokka().stacks.create(internalState.auth.organization, name, operations, options)
}

/**
 * Delete a stack
 *
 * @param {string} name
 *
 * @return {Promise}
 */
export function deleteStack(name) {
  return rokka()
    .stacks.delete(internalState.auth.organization, name)
    .then(() => {
      refreshStacks()
    })
}

/**
 * Set global alert
 *
 * @param {string} type           error|pending|success
 * @param {string} message
 * @param {number} [timeout=null] Timeout in ms
 */
export function setAlert(type, message, timeout = null) {
  updateState({
    alert: {
      type,
      message
    }
  })

  if (timeout) {
    setTimeout(() => {
      removeAlert()
    }, timeout)
  }
}

/**
 * Remove global alert
 */
export function removeAlert() {
  updateState({ alert: null })
}

/**
 * Subscribe to state updates
 *
 * @param {function} cb Listener
 *
 * @returns {function} Unsubscribe function
 */
export function subscribe(cb) {
  listener = cb
  return () => {
    listener = null
  }
}

/**
 * Clone stack
 *
 * @param {string} name
 * @param {object} operations
 * @param {object} options
 */
export function cloneStack(name, operations, options) {
  updateState({
    stackClone: {
      name: name,
      operations: operations,
      options: options
    }
  })
}

/**
 * Reset stack clone state
 *
 */
export function resetStackClone() {
  updateState({
    stackClone: {}
  })
}

// export state as readonly
const state = Object.assign({}, internalState)
export default state
