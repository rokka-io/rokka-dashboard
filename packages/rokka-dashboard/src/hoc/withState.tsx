import React, { ComponentType, useEffect, useState, FunctionComponent } from 'react';
import { AppState, currentState, subscribe, unsubscribe } from '../state';

type Reducer<S> = (state: AppState) => S;

/**
 * withState is a Higher-Order-Component which passes the global state (possibly reduced by the reducer function).
 */
export function withState<TInputProps, TOutputProps>(
  WrappedComponent: ComponentType<TOutputProps>,
  reducer: Reducer<TOutputProps>
): FunctionComponent<TInputProps> {
  return (props: TInputProps) => {
    const [compState, setCompState] = useState(reducer(currentState()));

    useEffect(() => {
      const updateState = (state: AppState) => {
        setCompState(reducer(state));
      };
      subscribe(updateState);
      return () => unsubscribe(updateState);
    }, []);

    return <WrappedComponent {...compState} {...props} />;
  };
}
