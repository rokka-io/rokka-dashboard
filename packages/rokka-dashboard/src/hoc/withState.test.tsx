import { mount } from 'enzyme';
import React from 'react';
import { testSubscriberCount, toggleSidebar } from '../state';
import { withState } from './withState';

describe('Higher-Order-Components', () => {
  describe('withState', () => {
    it('should forward the current global state as props to the component', () => {
      const Content = () => <div>some component</div>;
      const Comp = withState(Content);
      const component = mount(<Comp />);

      expect(component).toMatchSnapshot();

      expect(component.find(Content).props()).toStrictEqual({
        showSidebar: false
      });
    });
    it('should subscribe to state updates', () => {
      const Content = () => <div>some component</div>;
      const Comp = withState(Content);

      const component = mount(<Comp />);

      expect(component.find(Content).props()).toStrictEqual({
        showSidebar: false
      });

      toggleSidebar();

      // until setCompState is through and the component is re-rendered we need to wait
      // a setTimeout with 0 should help that wait time.
      setTimeout(() => {
        expect(component.find(Content).props()).toStrictEqual({
          showSidebar: true
        });
      }, 0);

      component.unmount();

      toggleSidebar();

      // after unmount unsubscribe() should be called and therefore the component should not receive any state
      // updates anymore.
      setTimeout(() => {
        expect(testSubscriberCount()).toBe(0);
        expect(component.find(Content).props()).toStrictEqual({
          showSidebar: true
        });
      }, 0);
    });
  });
});
