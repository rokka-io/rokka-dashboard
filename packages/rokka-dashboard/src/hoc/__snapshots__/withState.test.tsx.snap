// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Higher-Order-Components withState should forward the current global state as props to the component 1`] = `
<Component>
  <Content
    showSidebar={false}
  >
    <div>
      some component
    </div>
  </Content>
</Component>
`;
